ckan/model/base.py

```diff
@@ -1,6 +1,231 @@
+"""Base class for SQLAlchemy declaration-style models.
+"""
+
+
+class SessionMixin:
+    """Attach session to the model class.
+
+    Example:
+        class MyModel(SessionMixin, Base):
+            ...
+            @classmethod
+            def count(cls):
+                return cls.Session.query(cls).count()
+
+    """
+    Session = Session
+
+
+class DebugMixin:
+    """Defines __repr__ method that shows all the model's columns.
+
+    Example:
+        class MyModel(DebugMixin, Base):
+            name: str
+            ...
+
+        >>> MyModel(name="hello world")
+        <MyModel name=hello world>
+    """
+
+    def __repr__(self):
+        """Show model name and all the columns with their values.
+        """
+        output = u'<%s' % self.__class__.__name__
+        table: Any = orm.class_mapper(self.__class__).persist_selectable
+        for col in table.c:
+            try:
+                output += u' %s=%s' % (col.name, getattr(self, col.name))
+            except Exception as inst:
+                output += u' %s=%s' % (col.name, inst)
+
+        output += '>'
+        return output
+
+
+class NameMixin(SessionMixin):
+    """Add search and sorting by name to a model with `name` column.
+
+    Example:
+        class MyModel(NameMixin, Base):
+            name: str
+            ...
+
+        >>> MyModel.by_name("hello world")
+        None
+
+        >>> MyModel(name="hello world")
+        >>> MyModel.by_name("hello world")
+        <MyModel name=hello world>
+
+        >>> a_obj = MyModel(name="A")
+        >>> d_obj = MyModel(name="D")
+        >>> assert a_obj < d_obj
+    """
+
+    name: str
+
+    def __lt__(self, other: Self) -> bool:
+        return self.name < other.name
+
+    @classmethod
+    def by_name(
+            cls, name: Optional[str], autoflush: bool = True,
+            for_update: bool = False) -> Optional[Self]:
+        """Return first record with the given name.
+        """
+        q = cls.Session.query(cls).autoflush(autoflush).filter_by(name=name)
+        if for_update:
+            q = q.with_for_update()
+
+        return q.first()
+
+
+class TextSearchMixin:
+    """Provide base text-search functionality via LIKE SQL operator.
+
+    Text-search uses columns specified by `text_search_fields` property.
+
+    Example:
+        class MyModel(TextSearchMixin, Base):
+            text_search_fields = ["name]
+            name: str
+            ...
+
+        >>> MyModel(name="hello world")
+        >>> q = model.Session.query(MyModel)
+        >>> MyModel.text_search(q, "llo").all()
+        [<MyModel name=hello world>]
+    """
+    text_search_fields: list[str] = []
+
+    @classmethod
+    def text_search(cls, query: Any, term: str) -> Any:
+        """Add ILIKE filters to the query.
+        """
+        register = cls
+        make_like: Callable[
+            [Any, str], str] = lambda x, y: x.ilike('%' + y + '%')
+        q: Any = sa.null() | sa.null()
```

Comment:
`sa.null() | san.null()`?

Is this intended? 🤔

Reply:
Looks like a way to ensure that the query exists all the time because the next line begins a loop, that is going to append new clauses to the query.